docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,109 @@
+# Freeing objects and resources
+
+* Status: Propsed
+* Deciders: ?
+* Date: 2023-10-10
+
+Prevous discussion: [PR 1394](https://github.com/mozilla/uniffi-rs/issues/1394), [Issue 8](https://github.com/mozilla/uniffi-rs/issues/8)
+
+## Context and Problem Statement
+
+UniFFI must support objects that hold resources like database connections, file handles, network sockets, etc.  This is difficult, because objects like this are handled differently by different languages:
+
+* On Kotlin and other GC-based languages, the general advice is to explicitly close the resources and not rely on garbage collection to deal with them.
+* On Rust and other RAII-based languages, the general advice is to close the resources in a destructor.
+
+Currently, our Kotlin bindings try to address this problem for Objects by adding a `destroy()` method that
+manually drops the Rust value and implementing the `AutoCloseable` interface which allows users to
+ensure this is called via the `use` extension function.  However, this is not sufficient for several
+reasons:
+
+  * It is specific to Kotlin, but many languages face similar issues.
+  * All objects get this handling, when only some need it.
+  * This system is tied to decision to *not* support freeing UniFFI objects via GC-based systems
+    like `Cleaner`, which is not correct for some interfaces (see #1394)
+  * It only solves the issue in one direction, however UniFFI also needs to deal with the other
+    direction: a Kotlin resource-holding object can be passed to a Rust library via a callback
+    interface.  In this scenario, there's currently no mechanism for the Rust code to release the
+    Kotlin object's resource.
+
+We should come up with a general approach for resource-holding objects and apply it consistently
+across all languages.
+
+
+## Decision Drivers
+
+* If an object does not hold a resource, then we should be able to use the GC to free it.
+* The solution should work for both Rust objects passed to the foreign language and foreign objects
+  passed to Rust.
```

Comment:
IMO at least, but this isn't a hard requirement.  We could chose one option for objects and a different option for callabck interfaces (although [trait interfaces](https://github.com/mozilla/uniffi-rs/issues/1578) could get tricky).

Reply:
I'm actually against this, I think in contexts where RAII is available, we should use it. But maybe I'm misunderstanding something as I wasn't even aware that freeing of callback interface objects was ever a problem, and this ADR doesn't really provide any details as to *why* it's a problem (I would have thought dropping the `Box<dyn Trait>` would do the right thing,though now I'm wondering how it's even possible that Rust gets unique ownership over such an object, I guess it actually doesn't and the `Box` points to a GC-managed object?).